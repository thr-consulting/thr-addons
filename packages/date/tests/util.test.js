/* eslint-disable global-require */
import {LocalDate} from 'js-joda';
import moment from 'moment';
import {
	dateInit,
	transformDateToLocalDate,
	transformLocalDateToDate,
	transformMomentToLocalDate,
	transformLocalDateToMoment,
	transformMomentsToDate,
	transformDatesToMoment,
	transformLocalDatesToEpochInteger,
	transformEpochIntegerToLocalDate,
	mapEpochIntegerToLocalDates,
	transformObjectsToLocalDates,
	formatDate,
} from '../src/util';

jest.mock('react-widgets-moment');

expect.addSnapshotSerializer({
	test: v => moment.isMoment(v),
	print: v => v.toISOString(),
});
expect.addSnapshotSerializer({
	test: v => Object.prototype.toString.call(v) === '[object Date]',
	print: v => v.toJSON(),
});

describe('Date initialization', () => {
	it('should call the moment localizer', () => {
		const momentLocalizer = require('react-widgets-moment');
		dateInit();
		expect(momentLocalizer.mock.calls.length).toBe(1);
	});
});

describe('Date transforms', () => {
	it('should transform a JS date to a LocalDate', () => {
		const val = transformDateToLocalDate(new Date(2017, 5, 23, 6, 0, 0));
		expect(val).toBeInstanceOf(LocalDate);
		expect(val.dayOfMonth()).toBe(23);
		expect(val.monthValue()).toBe(6);
		expect(val.year()).toBe(2017);
	});

	it('should transform a LocalDate to a Date', () => {
		const val = transformLocalDateToDate(LocalDate.of(2017, 6, 23));
		expect(val).toBeInstanceOf(Date);
		expect(val.getDate()).toBe(23);
		expect(val.getMonth()).toBe(5);
		expect(val.getFullYear()).toBe(2017);
	});

	it('should transform a Moment to a LocalDate', () => {
		const val = transformMomentToLocalDate(moment([2017, 5, 23, 6, 0, 0]));
		expect(val).toBeInstanceOf(LocalDate);
		expect(val.dayOfMonth()).toBe(23);
		expect(val.monthValue()).toBe(6);
		expect(val.year()).toBe(2017);
	});

	it('should transform a LocalDate to a Moment', () => {
		const val = transformLocalDateToMoment(LocalDate.of(2017, 6, 23));
		expect(val).toBeInstanceOf(moment);
		expect(val.date()).toBe(23);
		expect(val.month()).toBe(5);
		expect(val.year()).toBe(2017);
	});

	it('should transform a single JS Date to a single Moment', () => {
		const val = transformDatesToMoment(new Date(2017, 5, 23, 6, 0, 0));
		expect(val).toBeInstanceOf(moment);
		expect(val.date()).toBe(23);
		expect(val.month()).toBe(5);
		expect(val.year()).toBe(2017);
	});

	it('should transform an array of Moments to an array of JS Dates', () => {
		const val = transformMomentsToDate([
			moment([2017, 5, 23, 6, 0, 0]),
			moment([2010, 10, 5, 10, 5, 6]),
		]);
		expect(val[0]).toBeInstanceOf(Date);
		expect(val[1]).toBeInstanceOf(Date);
		expect(val[0].getDate()).toBe(23);
		expect(val[0].getMonth()).toBe(5);
		expect(val[0].getFullYear()).toBe(2017);
		expect(val[1].getDate()).toBe(5);
		expect(val[1].getMonth()).toBe(10);
		expect(val[1].getFullYear()).toBe(2010);
	});

	it('should transform an array of JS Dates to an array of Moments', () => {
		const val = transformDatesToMoment([
			new Date(2017, 5, 23, 6, 0, 0),
			new Date(2010, 10, 5, 10, 5, 6),
		]);
		expect(val[0]).toBeInstanceOf(moment);
		expect(val[1]).toBeInstanceOf(moment);
		expect(val[0].date()).toBe(23);
		expect(val[0].month()).toBe(5);
		expect(val[0].year()).toBe(2017);
		expect(val[1].date()).toBe(5);
		expect(val[1].month()).toBe(10);
		expect(val[1].year()).toBe(2010);
	});

	it('should transform an object with Moment fields to the same object with JS Dates', () => {
		const val = transformMomentsToDate({
			field1: moment([2017, 5, 23, 6, 0, 0]),
			field2: moment([2010, 10, 5, 10, 5, 6]),
		});
		expect(val.field1).toBeInstanceOf(Date);
		expect(val.field2).toBeInstanceOf(Date);
		expect(val.field1.getDate()).toBe(23);
		expect(val.field1.getMonth()).toBe(5);
		expect(val.field1.getFullYear()).toBe(2017);
		expect(val.field2.getDate()).toBe(5);
		expect(val.field2.getMonth()).toBe(10);
		expect(val.field2.getFullYear()).toBe(2010);
	});

	it('should transform an object with JS Date fields to the same object with Moments', () => {
		const val = transformDatesToMoment({
			field1: new Date(2017, 5, 23, 6, 0, 0),
			field2: new Date(2010, 10, 5, 10, 5, 6),
		});
		expect(val.field1).toBeInstanceOf(moment);
		expect(val.field2).toBeInstanceOf(moment);
		expect(val.field1.date()).toBe(23);
		expect(val.field1.month()).toBe(5);
		expect(val.field1.year()).toBe(2017);
		expect(val.field2.date()).toBe(5);
		expect(val.field2.month()).toBe(10);
		expect(val.field2.year()).toBe(2010);
	});

	it('should transform an object with LocalDates to the same object with Epoch Integers', () => {
		const val = transformLocalDatesToEpochInteger({
			field1: LocalDate.ofEpochDay(17400),
			field2: LocalDate.ofEpochDay(13000),
		});
		expect(val.field1).toBe(17400);
		expect(val.field2).toBe(13000);
	});

	it('should transform an epoch integer into a LocalDate', () => {
		const val = transformEpochIntegerToLocalDate(17400);
		expect(val.dayOfMonth()).toBe(22);
		expect(val.monthValue()).toBe(8);
		expect(val.year()).toBe(2017);
	});

	it('should map epoch integers into LocalDates', () => {
		const src1 = {
			field1: 17400,
			field2: {subField2: 17400, subField3: 13000},
			field3: [17400, 13000],
			field4: [17400, 13000],
			field5: [{a: 17400, b: 17400}, {a: 13000, b: 13000}],
			field6: [{a: [{b: [17400, 13000]}]}],
		};
		const val = mapEpochIntegerToLocalDates(src1, [
			'field1',
			'field2.subField2',
			'field3',
			'field5.a',
			'field6.a.b',
		]);
		expect(val.field1).toBeInstanceOf(LocalDate);
		expect(val).toMatchSnapshot();
	});

	it('should transform objects to LocalDates', () => {
		const val = transformObjectsToLocalDates({
			field1: LocalDate.ofEpochDay(17400),
			field2: [LocalDate.ofEpochDay(13000), LocalDate.ofEpochDay(1)],
		});
		expect(val.field1).toBeInstanceOf(LocalDate);
		expect(val.field2[0]).toBeInstanceOf(LocalDate);
		expect(val).toMatchSnapshot();
	});
});

describe('Formatting', () => {
	it('should format with defaults', () => {
		expect(formatDate(moment([2017, 5, 23, 6, 0, 0]))).toMatchSnapshot();
	});
	it('should format with options', () => {
		expect(formatDate(moment([2017, 5, 23, 6, 0, 0]), {type: 'short', time: true, date: true})).toMatchSnapshot();
		expect(formatDate(moment([2017, 5, 23, 6, 0, 0]), {type: 'medium', time: true, date: true})).toMatchSnapshot();
		expect(formatDate(moment([2017, 5, 23, 6, 0, 0]), {type: 'long', time: true, date: true})).toMatchSnapshot();
	});
	it('should use a custom date format', () => {
		expect(formatDate(moment([2017, 5, 23, 6, 0, 0]), {format: 'MMMM_D_YYYY_h_mm_a'})).toMatchSnapshot();
	});
});
