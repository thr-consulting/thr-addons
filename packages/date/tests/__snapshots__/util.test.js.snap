// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date transforms should map epoch integers into LocalDates 1`] = `
Object {
  "field1": "2017-08-22",
  "field2": Object {
    "subField2": "2017-08-22",
    "subField3": 13000,
  },
  "field3": Array [
    "2017-08-22",
    "2005-08-05",
  ],
  "field4": Array [
    17400,
    13000,
  ],
  "field5": Array [
    Object {
      "a": "2017-08-22",
      "b": 17400,
    },
    Object {
      "a": "2005-08-05",
      "b": 13000,
    },
  ],
  "field6": Array [
    Object {
      "a": Array [
        Object {
          "b": Array [
            "2017-08-22",
            "2005-08-05",
          ],
        },
      ],
    },
  ],
}
`;

exports[`Date transforms should transform objects to LocalDates 1`] = `
Object {
  "field1": "2017-08-22",
  "field2": Array [
    "2005-08-05",
    "1970-01-02",
  ],
}
`;

exports[`Formatting should format with defaults 1`] = `"6/23/2017"`;

exports[`Formatting should format with options 1`] = `"6/23/2017 6:00 am"`;

exports[`Formatting should format with options 2`] = `"Jun 23, 2017 6:00 am"`;

exports[`Formatting should format with options 3`] = `"June 23, 2017 6:00 am"`;

exports[`Formatting should use a custom date format 1`] = `"June_23_2017_6_00_am"`;
